Framework/PythonInterface/plugins/algorithms/IntegratePeaksShoeboxTOF.py

```diff
@@ -0,0 +1,687 @@
+                    result.plot_integrated_peak(fig, axes, LogNorm, Rectangle)
+                    pdf.savefig(fig)
+                    close(fig)
+    except OSError:
+        raise RuntimeError(
+            f"OutputFile ({output_file}) could not be opened - please check it is not open by "
+            f"another programme and that the user has permission to write to that directory."
+        )
+
+
+def calc_angle_between_peaks(pk1, pk2):
+    return abs(pk1.getQLabFrame().angle(pk2.getQLabFrame()))
+
+
+def get_and_clip_data_arrays(ws, peak_data, pk_tof, ispec, kernel, nshoebox):
+    x, y, esq, ispecs = peak_data.get_data_arrays()  # 3d arrays [rows x cols x tof]
+    x = x[peak_data.irow, peak_data.icol, :]  # take x at peak centre, should be same for all detectors
+    # crop data array to TOF region of peak using shoebox dimension
+    itof = ws.yIndexOfX(pk_tof, ispec)  # need index in y now (note x values are points even if were edges)
+    tof_slice = slice(
+        int(np.clip(itof - nshoebox * kernel.shape[-1] // 2, a_min=0, a_max=len(x))),
+        int(np.clip(itof + nshoebox * kernel.shape[-1] // 2, a_min=0, a_max=len(x))),
+    )
+    x = x[tof_slice]
+    y = y[:, :, tof_slice]
+    esq = esq[:, :, tof_slice]
+    return x, y, esq, ispecs
+
+
+def convolve_shoebox(y, esq, kernel, mode="same"):
+    yconv = convolve(y, kernel, mode=mode)
+    with np.errstate(divide="ignore", invalid="ignore"):
+        econv = np.sqrt(convolve(esq, kernel**2, mode=mode))
+        intens_over_sig = yconv / econv
+    intens_over_sig[~np.isfinite(intens_over_sig)] = 0
+    intens_over_sig = uniform_filter(intens_over_sig, size=len(y.shape))
+    # zero edges where convolution is invalid
+    if mode == "same":
+        edge_mask = np.ones(intens_over_sig.shape, dtype=bool)
+        center_slice = tuple(slice(nedge, -nedge) for nedge in (np.asarray(kernel.shape) - 1) // 2)
+        edge_mask[center_slice] = False
+        intens_over_sig[edge_mask] = 0
+    return intens_over_sig
+
+
+def find_nearest_peak_in_data_window(data, ispecs, x, ws, peaks, ipk, irow, icol, ix, threshold=2):
+    ipks_near, ispecs_peaks = find_ipks_in_window(ws, peaks, ispecs, ipk, tof_min=x.min(), tof_max=x.max())
+    if len(ipks_near) > 0:
+        tofs = peaks.column("TOF")
+        # get position of nearby peaks in data array in fractional coords
+        shape = np.array(data.shape)
+        pos_near = [
+            np.r_[np.argwhere(ispecs == ispecs_peaks[ii])[0], np.argmin(abs(x - tofs[ipk_near]))] / shape
+            for ii, ipk_near in enumerate(ipks_near)
+        ]
+        # sort data in descending order and select strongest data point nearest to pk (in fractional coordinates)
+        isort = list(zip(*np.unravel_index(np.argsort(-data, axis=None), data.shape)))
+        pk_pos = np.array([irow, icol, ix]) / np.array(data.shape)
+        imax_nearest = None
+        for ibin in isort:
```

Comment:
`isort` can be quite a large list. Can you select only the appropiate data with a [data>threshold] mask, so that `isort` only contains the indices of data for which data>threshold? This way you don't need to check for that condition in the `for` loop and it reduces the loop count drastically.

Reply:
Yes this can be done, I'd need to change the logic a bit to work if `isort` was empty but it's a good suggestion thank you!

Reply:
Looking at this I need to convert the index from the flattened array to the ndarray, which is why I don't do e.g. `np.argsort(-data[data>2], axis=None)`. However, I can still reduce the loop count by adding another break as soon as the data goes below the threshold as the sequence is sorted.